### ISSUE TITLE
capture and validate backed up files checksums

### ISSUE THREAD

**Issue by ifedorenko** [restic/restic#1620]
As a (paranoid) user, I want `backup` to capture SHA256 checksum of files I back up. I also want `restore` to fail if the file checksum after restore does not match the checksum captured during backup. Likewise, I want `check` to have an option to check file data checksum matches the checksum captured during backup.

**Comment by rawtaz** [comment]
For `backup`, the blobs in which your data in the repository are stored when backing up *are* already saved with a name that corresponds to the checksum of the data they hold. It does use SHA256 for this.

The `restore` command *will* alert you if upon restoring a blob, the checksum of the data it derives/gets from it no longer corresponds to the checksum that was stored as part of the backup.

The `check` command does some basic checks, but if you pass it the `--read-data` option, it will make sure to check these things for you (by specifically read all the data and check that it still matches the expected checksums.

Is there anything in the above which does not cover the things you want? I'm not sure it's clear what this issue is really about.

It would be great if in the future you use the issue template, as it helps structuring the content of the issue in a way such that we get a clearer picture of it. And if it's a more general "discussion-like" topic you want to bring up, the [forum](https://forum.restic.net) is a great place for that. Thanks!

**Comment by ifedorenko** [comment]
Say I backup a 10MB file named "PICTURE.JPEG". This file will be split in ~10 blobs and stored in 2+ pack files in the repository. While restic does checksum and validate individual blobs and pack files used to store the file data in the repository (as well as other internal data structures), it does not currently checksum/validate the file as a whole, and this is what I am asking in this user story.

**Comment by rawtaz** [comment]
Right, understood. So, what problem does this solve? If the verification of those ten different parts of the file is good, then what you want to check is basically that restic restored them in the right order, or that there is no hardware issue causing the restored file to be corrupt?

**Comment by ifedorenko** [comment]
Right, I want to be able to check if restic restored parts in the right order and didn't lose any parts. I also want to manually check if files I have locally match files in the backup.

**Comment by jdeng** [comment]
Any update on this issue? I think having SHA256 checksums for the files (e.g., stored as an attribute within the node data structure) will be handy. Applications can use this field to detect duplicated files and to validate the files independently etc. This field can be optional and turned off by default.

One use case is to scan all the trees and build a SHA256 map of all the files without reading the blobs.

**Comment by fd0** [comment]
That's one way to implement this, yes. Thanks for the proposal.

A few random thoughts from me:

* I think it's not a good idea to support several algorithms. When we add this we should decide which hash it should be, then we can amend the design document and be done with it. SHA256 would be my choice, since we're already using it everywhere.
 * Before this can be implemented we need to decide what to do with the user interface and the settings for this option. I think we should (at most) have an on/off switch, nothing more. If it's off by default, nobody will turn it on. If it's on by default, people can turn it off if it causes too much CPU load. I'm leaning towards activating this by default, with an off switch.
 * Adding a new (optional) field in the repository data structures is a non-breaking change, we just need to agree on the details. A backup with a new version of restic adding this field may re-write all the metadata.
 * Whenever code accesses the new field, there must be a graceful fallback if the field is unset.
 * It'd be great if the PR adding this field (or a follow-up PR proposed shortly after) adds functionality which uses the additional information, so we have something we can show users.

**Comment by jdeng** [comment]
Thanks for the feedback! The reason I think multiple hash algorithms should be supported is that users may have chosen their own file hashing algorithms and switching is probably not convenient.  For example, Google Drive and some Smugmug are providing MD5 checksums, and git is using SHA1.  If a user want to build a HTTP frontend on top of a restic backup (similar to the dump command) and using the same hashing algorithm (to check duplications and/or look up content with a database where they stored the file hashes), I think having such option is important and we can make SHA256 the default.

Agreed on other points.  The code changes are meant for illustrating the idea and the user interface needs more thinking. It seems to me like a secondary index (file id to node) is probably the ideal outcome but requires more work and I haven't look into how to make the changes. A full scan of the index to get a file by ID seems inefficient but doable (and use the same dump command). Other options can be building an external index with mapping of file id to tree id which can accelerate the lookup but a little unintuitive. Any suggestions?

**Comment by cfbao** [comment]
I'd also like to have file checksums captured for the purpose of having a better file version history.
Metadata like mtime, ctime can often be unreliable at telling whether a file actually changed. This limits the usefulness of the `find` command. Having file level checksums/hashes can solve this problem.

More discussions in this forum thread: https://forum.restic.net/t/per-file-version-history/929

**Comment by Sina1910** [comment]
I would be glad to find this feature implemented soon.

Advanatages relevant for me would be (some copied from above):
- check mechanism when restoring files
- ability to manually check if files I have locally match files in the backup
- basis to detect duplicated files
- basis for building a map with checksumsof all the files in a repo
- better file version history (Metadata like mtime, ctime often not reliable)
- identify renamed/relocated files

**Comment by deliciouslytyped** [comment]
I would like to have sha256 hashes or the like available so that I can check for duplicate files that I don't want to sync into my respository. Storing these files is not a storage (due to restic deduplication) issue, it's an organization issue.

The only manner of workaround I could think of is hashing the files in a mount, but for a multi-terabyte remote repository that's not an option - and this is inefficient anyway.

Is there any reasonable workaround that doesn't involve trying to hash my data over a network connection?

**Comment by johnmaguire** [comment]
I'm not sure what the implementation would look like, but another interesting idea would be to have the option to store parity files. For example, Gareth George (author of the backrest UI) talks about using par2 in his archives here: https://gareth.page/post/2023-11-02-data-immortality/#erasure-codes-for-great-good-and-the-benefits-of-par2-archives

**Comment by MichaelEischer** [comment]
@johnmaguire The parity discussion belongs into https://github.com/restic/restic/issues/804 .

**Comment by theosaveliev** [comment]
>Before this can be implemented we need to decide what to do with the user interface

I really need a `diff` command that compares local files against the backup. Just like git status.

The use-case is the following:
- Run `restic diff --local` to check local filesystem against the backup
- Find out whether the local files are corrupted
- Do not create a snapshot

Right now, I do this as follows:
1. Create a snapshot (slow and takes network traffic)
2. Run `restic diff` to see what's changed between 2 latest snapshots
3. Remove the snapshot
